### PR TITLE
Add support for many new devices

### PR DESCRIPTION
**Pull request by daymondm**
Adds support for: 

* light switches
* dimmer switches
* arming/disarming panel
* garage door
* camera motion detectors

Also, setTimeout function is fixed.

### CODE REVIEW

**Comment by mceres** [comment]
Getting error for a missing module... i am guessing i need to install that module for the camera feed?

image

**Comment by mceres** [comment]
you are probably working on the camera feed, i commented out the following line to load vivint for now
var FFMPEG = require('../../homebridge-camera-ffmpeg/ffmpeg.js').FFMPEG

also changed the code on the dimmer switch to test the behavior i had and seems to be working a little better, i do see a toggle in homekit but now it goes back to off instead of turning on the homekit switch and to a random brightness

setSwitchCurrentValue(targetState, next) {
                        if (targetState) {
                                // turn switch on
                                vivintApi.putDevice('switches', this.id, {
                                                **val: 100,**
                                                _id: this.id
                                        })
                                        .then(
                                                (success) => next(),
                                                (failure) => {
                                                        log("failure " + failure)
                                                        next(failure)
                                                })
                        } else {
                                // turn switch off
                                vivintApi.putDevice('switches', this.id, {
                                                **val: 0,**
                                                _id: this.id
                                        })
                                        .then(
                                                (success) => next(),
                                                (failure) => {
                                                        log("failure " + failure)
                                                        next(failure)
                                                })
                        }
                }

File: lib/vivint_api.js [hunk 1]
```diff
@@ -63,12 +63,31 @@ function VivintApiModule(config, log) {
     }
 
     putDevice(category, id, data) {
+      //log("(" + data.id + ") Body: " + JSON.stringify(data))
```
Comment (timcharper) [inline lib/vivint_api.js]: please delete these comments

File: README.md [hunk 2]
```diff
@@ -36,5 +44,4 @@ Configuration options overview:
 * **password**
 * **apiLoginRefreshSecs** - How often should Vivint Homebridge renew the session token? The token that Vivint provides when authenticating will expire. Also, when this renewal occurs, the plugin requests another snapshot. The event stream can sometimes fail to report device state appropriately and events can come out of order with the snapshot, or updates can be missed entirely. The occasional snapshot retrieval will auto-correct any such errors. Avoid setting this any more frequent that 10 minutes.
 * **motionDetectedOccupancySensorMins** - Homebridge-vivint will create occupancy sensors for motion sensors that will stay active for X minutes after a motion event is detected. This value configures for how long that occupancy sensor will stay active if no further motion events are detected. Note: Vivint's reporting of motion events over the event stream can be a little inconsistent, at times. As a recommendation, don't plan on creating Homekit automations that respond to Vivint motion events.
-
-
+* **lowbatterylevel** - How low the battery level should be before the plugin reports to HomeKit that the battery is low.
```
Comment (timcharper) [inline README.md:46]: great docs!

File: README.md [hunk 3]
```diff
@@ -15,6 +15,13 @@ Currently, the following items are supported:
 * Thermostat
 * Motion sensors
```
Comment (timcharper) [inline README.md]: remove spurious double-space

File: lib/device_set.js [hunk 4]
```diff
@@ -1,539 +1,1106 @@
...
-        return Characteristic.ContactSensorState.CONTACT_DETECTED
-    }
-
-    notify() {
-      this.service.getCharacteristic(Characteristic.ContactSensorState)
-        .updateValue(this.contactSensorValue())
-    }
-  }
-
-  extend(ContactSensor, {
```
Comment (timcharper) [inline lib/device_set.js]: Can you try and match the existing code styling? (is there a JavaScript code formatter we can use?)

I don't really care how it is formatted, so much that it is consistent and that unrelated lines aren't changed.

**Review by timcharper — CHANGES_REQUESTED** [review]

**Comment by timcharper** [comment]
maybe we can use https://www.npmjs.com/package/js-beautify ? We should apply a change introducing js-beautify first, (or something similar) and then redo this change with it run.

Again, I don't care about the settings, I just care about consistency :) 4 spaces, 2 spaces, ¯\_(ツ)_/¯ just consistent.

**Comment by daymondm** [comment]
@mceres try this?

```
  class DimmerSwitch extends Device {
    constructor(accessory) {
      super(accessory)

      this.service = accessory.getServiceByUUIDAndSubType(Service.Lightbulb)

      this.service
        .getCharacteristic(Characteristic.On)
        .on('get', (next) => next(null, this.switchCurrentValue()))
        .on('set', this.setSwitchCurrentValue.bind(this))

      this.service
        .getCharacteristic(Characteristic.Brightness)
        .on('get', (next) => next(null, this.switchBrightnessValue()))
        .on('set', this.setBrightnessValue.bind(this))

      this.last_val_s = null
      this.last_val = null
    }

    switchCurrentValue() {
      return this.data.s
    }

    setSwitchCurrentValue(targetState, next) {
      if (targetState) {
        // turn switch on
        vivintApi.putDevice('switches', this.id, {
            s: true,
            _id: this.id
          })
          .then(
            (success) => next(),
            (failure) => {
              log("failure " + failure)
              next(failure)
            })
      } else {
        // turn switch off
        vivintApi.putDevice('switches', this.id, {
            s: false,
            _id: this.id
          })
          .then(
            (success) => next(),
            (failure) => {
              log("failure " + failure)
              next(failure)
            })
      }
    }

    switchBrightnessValue() {
      return this.data.val
    }

    setBrightnessValue(targetState, next) {
      vivintApi.putDevice('switches', this.id, {
          val: targetState,
          _id: this.id
        })
        .then(
          (success) => next(),
          (failure) => {
            log("failure " + failure)
            next(failure)
          })
    }

    notify() {
      // vivint will typically send two notifications

      if (this.service) {
        if (this.data.val !== this.last_val) {
          // dimmer level has changed, update HomeKit
          this.last_val = this.data.val
          this.service
            .getCharacteristic(Characteristic.Brightness)
            .updateValue(this.data.val)
        }

        if (this.data.s !== this.last_val_s) {
          // on/off state has changed, update HomeKit
          this.last_val_s = this.data.s
          this.service
            .getCharacteristic(Characteristic.On)
            .updateValue(this.switchCurrentValue())
        }
      }
    }
  }
```

I'm also curious if vivint will accept two arguments... if the above doesn't work, try:

```
        // turn switch on
        vivintApi.putDevice('switches', this.id, {
            s: true,
            val: 100,
            _id: this.id
          })
          .then(
            (success) => next(),
            (failure) => {
              log("failure " + failure)
              next(failure)
            })
```

File: lib/device_set.js (hunk 4 above)
Comment (daymondm) [reply lib/device_set.js]: I had ran it through a beautifier at https://beautifier.io ... sorry I've used a couple different text editors and landed on atom at the moment. Apparently I had made some changes and forgot to run it through the beautifier again! Going to see if I can resolve the dimmer issue first. You asked if I meant to close the PR... Do I need to close the PR and reopen a new one? or will it update if I upload the file again?

**Comment by mceres** [comment]
@daymondm i tried bot codes and still same behavior.. using only the val:0 and val:100 seems to work better.. maybe is an issue with vivint on dimmer switches and sends both values for val and s instead of using one or the other.

**Comment by profbass** [comment]
I am very excited for these updates. Thanks for putting them together guys.

**Comment by jlhunter2008** [comment]
@timcharper any timetable when you’ll make these changes live?

**Comment by brianpham** [comment]
Excited for this change as well. Let me know if I can help with anything

**Review by timcharper — APPROVED** [review]

**Comment by timcharper** [comment]
Great work @daymondm ! I've merged it. Thanks for your contribution!
